alpenhorn/auto_import.py

```diff
@@ -1,323 +1,357 @@
+    # Strip node root, if path is absolute
+    if path.is_absolute():
+        try:
+            path = path.relative_to(node.db.root)
+        except ValueError:
+            # Not rooted under node.root
+            log.warning(
+                f"skipping import of {path}: not rooted under node {node.db.root}"
             )
             return
-    else:
-        file_path = os.path.join(node.root, file_path)
-    abspath = os.path.normpath(file_path)
 
-    # Skip requests to import a directory. Again these are occasionally sent by the watchdog
-    if os.path.isdir(file_path):
-        log.debug('Path to import "%s" is a directory. Skipping...', file_path)
+    # Ignore the node info file
+    if str(path) == "ALPENHORN_NODE":
+        log.debug("ignoring ALPENHORN_NODE file during import")
         return
 
-    relpath = os.path.relpath(abspath, node.root)
+    # New import task.  It's better to do this in a worker
+    # because the worker pool will take care of DB connection loss.
+    Task(
+        func=_import_file,
+        queue=queue,
+        key=node.name,
+        args=(node, path),
+        name=f"Import {path} on {node.name}",
+        # If the job fails due to DB connection loss, re-start the
+        # task because unlike tasks made in the main loop, we're
+        # never going to revisit this.
+        requeue=True,
+    )
+
+
+def _import_file(task: Task, node: StorageNode, path: pathlib.PurePath) -> None:
+    """Import `path` on `node` into the DB.  This is run by a worker.
+
+    Parameters
+    ----------
+    task : task.Task
+        The task running this function
+    node : storage.StorageNode
+        The node we are processing.
+    path : pathlib.PurePath
+        The path should be relative to `node.root`
+    """
+
+    log.debug(f'Considering "{path}" for import to node {node.name}.')
+
+    # Wait for file to become ready
+    while not node.io.ready_path(path):
+        log.debug(
+            f'Path "{path}" not ready for I/O during import.  Waiting 60 seconds.'
+        )
+        yield 60  # wait 60 seconds and re-queue
 
     # Skip the file if there is still a lock on it.
-    dir_name, base_name = os.path.split(abspath)
-    if os.path.isfile(os.path.join(dir_name, ".%s.lock" % base_name)):
-        log.debug('Skipping "%s", which is locked by ch_master.py.', file_path)
+    if node.io.locked(path):
+        log.debug(f'Skipping "{path}": locked.')
         return
 
     # Step through the detection extensions to find one that's willing
     # to handle this file
     for detector in extensions.import_detection():
-        acq_name, callback = detector(abspath)
+        acq_name, callback = detector(path, node)
         if acq_name is not None:
             break
+    else:
+        acq_name = None
 
     # Was detection successful?
     if acq_name is None:
-        log.info('Skipping non-acquisition path "%s".', file_path)
+        log.info(f"Skipping non-acquisition path: {path}")
```

Comment:
Maybe this changes in a future PR, but couldn't this just move directly into the `else` statement, since it returns right away?

Reply:
:+1: It's a remnant of one of the many iterations of this PR.